src/raft.c

```diff
@@ -177,6 +179,62 @@ void RaftExecuteCommandArray(RedisModuleCtx *ctx,
     }
 }
 
+static void lockKeys(RedisRaftCtx *rr, raft_entry_t *entry)
+{
+    assert(entry->type == RAFT_LOGTYPE_LOCK_KEYS);
+
+    RaftReq *req = entry->user_data;
+
+    /* FIXME: can optimize this for leader by getting it out of the req, keeping code simple for now */
+    size_t num_keys;
+    RedisModuleString ** keys = RaftRedisLockKeysDeserialize(entry->data, entry->data_len, &num_keys);
+
+    for (size_t i = 0; i < num_keys; i++) {
+        size_t str_len;
+        const char * str = RedisModule_StringPtrLen(keys[i], &str_len);
+        LOG_WARNING("locking %.*s", (int) str_len, str);
+        RedisModule_DictSet(rr->locked_keys, keys[i], NULL);
+        RedisModule_FreeString(rr->ctx, keys[i]);
+    }
+
+    RedisModule_Free(keys);
+
+    if (req) {
+        /* currently, we just return, but eventually this is where should kick off migration of the keys */
+        RedisModule_ReplyWithSimpleString(req->ctx, "OK");
+        RaftReqFree(req);
+//        MigrateKeys(rr, req);
+    }
+}
+
+static void unlockDeleteKeys(RedisRaftCtx *rr, raft_entry_t *entry)
+{
+    assert(entry->type == RAFT_LOGTYPE_DELETE_UNLOCK_KEYS);
+
+    RaftReq *req = entry->user_data;
+
+    size_t num_keys;
+    RedisModuleString ** keys;
+
+    keys = RaftRedisLockKeysDeserialize(entry->data, entry->data_len, &num_keys);
+
+    enterRedisModuleCall();
+    RedisModuleCallReply *reply = RedisModule_Call(redis_raft.ctx, "del", "v", keys, num_keys);
```

Comment:
Maybe we can assert() here, just to verify command has been executed.

Reply:
yes, that's probably a good idea.  it shouldn't be able to error, but it can fail to execute and get a null response.